### PR TITLE
[GOBBLIN-1641] Add meter for sla exceeded flows

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1641

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Currently the dag manager does not have metrics around how many flows are being exceeded per executor, so we should track this.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 1]
```diff
@@ -457,7 +460,8 @@ public static class DagManagerThread implements Runnable {
     DagManagerThread(JobStatusRetriever jobStatusRetriever, DagStateStore dagStateStore, DagStateStore failedDagStateStore,
         BlockingQueue<Dag<JobExecutionPlan>> queue, BlockingQueue<String> cancelQueue, BlockingQueue<String> resumeQueue,
         boolean instrumentationEnabled, Set<String> failedDagIds, ContextAwareMeter allSuccessfulMeter,
-        ContextAwareMeter allFailedMeter, Long defaultJobStartSla, UserQuotaManager quotaManager, int dagMangerThreadId) {
+        ContextAwareMeter allFailedMeter, Map<String, ContextAwareMeter> startSlaExceededMeters,
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:463]: not mandatory now, but seeing there are already three `CAMeter`-related params makes me wonder whether a higher-level of abstraction might be in order (to collect them together).

perhaps wait and see?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 2]
```diff
@@ -1098,6 +1105,19 @@ private ContextAwareMeter getGroupMeterForDag(String dagId, String meterName, Ma
...
 
+    /**
+     * Used to track metrics for different specExecutors to detect issues with the specExecutor itself
+     * @param dagNode
+     * @param meterName
+     * @param meterMap
+     * @return
+     */
+    private ContextAwareMeter getExecutorMeterForDag(DagNode<JobExecutionPlan> dagNode, String meterName, Map<String, ContextAwareMeter> meterMap) {
+      String executorName = dagNode.getValue().getSpecExecutor().getUri().toString();
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1116]: if we lack a `DagManagerUtils.getExecutorName`, should we add?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 3]
```diff
@@ -1098,6 +1105,19 @@ private ContextAwareMeter getGroupMeterForDag(String dagId, String meterName, Ma
...
+      return meterMap.computeIfAbsent(executorName,
+          executorUri -> metricContext.contextAwareMeter(MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX, executorUri, meterName)));
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1118]: to turn this on its head, the map you're carrying around and passing into here is basically just function memoization for this `computeIfAbsent` call.

up to you, but I might raise the level of abstraction by presenting this as a cache of `ContextAwareMeter`s that delegates read-through (initialization) to the `MetricRegistry`.

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java [hunk 4]
```diff
@@ -736,6 +736,9 @@ public void testJobStartSLAKilledDag() throws URISyntaxException, IOException {
     Assert.assertEquals(this.dagToJobs.size(), 1);
     Assert.assertTrue(this.dags.containsKey(dagId1));
 
+    String slakilledMeterName = MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX, "job0", ServiceMetricNames.START_SLA_EXCEEDED_FLOWS_METER);
+    Assert.assertEquals(metricContext.getParent().get().getMeters().get(slakilledMeterName).getCount(), 1);
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagManagerTest.java:740]: would it be worth verifying that meters are per-executor, not a single aggregation?

**Review by phet — COMMENTED** [review]
having this per-executor will be very helpful to delineating and detecting outages (albeit with a lag of the start SLA (deadline))

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 5]
```diff
@@ -357,6 +357,8 @@ public synchronized void setActive(boolean active) {
 
         ContextAwareMeter allSuccessfulMeter = null;
         ContextAwareMeter allFailedMeter = null;
+        // TODO: create a map of RatioGauges for success/failed per executor, will require preprocessing of available executors
+        Map<String, ContextAwareMeter> startSlaExceededMeters = Maps.newConcurrentMap();
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:361]: a. I was thinking of a more robust metrics accounting that could do RatioGauges, which is an abstract class https://www.javadoc.io/static/io.dropwizard.metrics/metrics-core/3.1.2/com/codahale/metrics/RatioGauge.html that takes in 2 meters. I was thinking that if we were to also report ratios, then they need a reference to both the corresponding meters as well? Haven't thought it out all the way.

b. I was thinking with ratios we'd have percentage of successful/failed jobs

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java (hunk 2 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1116]: I can add that, that's a good point :)

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java (hunk 3 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1118]: Ah yeah that's a good point it is essentially a cache 😓 . I think we can leave it as is for now but when we refactor the metrics out of the DagManager this should have a different interface along with the other metrics for some consolidation/consistency.

**Review by ZihanLi58 — APPROVED** [review]
+1

**Review by phet — APPROVED** [review]
I like the robust testing!

**Pull request merged**